nucliadb_core/src/query_language/mod.rs

```diff
@@ -0,0 +1,442 @@
+        let BooleanExpression::Literal(literal) = operand else {
+            #[rustfmt::skip] return Err(node_error!(
+                "Nested expressions can not be splitted"
+            ));
+        };
+        if context.field_labels.contains(&literal) {
+            fields_only.push(BooleanExpression::Literal(literal));
+        } else {
+            paragraphs_only.push(BooleanExpression::Literal(literal))
+        }
+    }
+
+    Ok(ExpressionSplit {
+        fields_only: if fields_only.len() == 1 {
+            fields_only.pop().unwrap()
+        } else {
+            BooleanExpression::Operation(BooleanOperation {
+                operator: Operator::And,
+                operands: fields_only,
+            })
+        },
+        paragraphs_only: if paragraphs_only.len() == 1 {
+            paragraphs_only.pop().unwrap()
+        } else {
+            BooleanExpression::Operation(BooleanOperation {
+                operator: Operator::And,
+                operands: paragraphs_only,
+            })
+        },
+    })
+}
+
+#[derive(Default)]
+pub struct QueryAnalysis {
+    pub prefilter_query: Option<BooleanExpression>,
+    pub search_query: Option<BooleanExpression>,
+}
+
+pub fn translate(query: String, context: QueryContext) -> NodeResult<QueryAnalysis> {
+    let Json::Object(subexpressions) = serde_json::from_str(&query)? else {
+        #[rustfmt::skip] return Err(node_error!(
+            "Unexpected query format {query}"
+        ));
+    };
+
+    let mut prefilter_queries = Vec::new();
+    let mut search_queries = Vec::new();
+
+    for (root_id, expression) in subexpressions {
+        let as_json_expression = serde_json::json!({ root_id: expression });
+        let translation_report = translate_json(as_json_expression, &context)?;
+        if translation_report.has_field_labels && translation_report.has_paragraph_labels {
+            let splitted = split_mixed(translation_report.expression, &context)?;
+            prefilter_queries.push(splitted.fields_only);
+            search_queries.push(splitted.paragraphs_only);
+        } else if translation_report.is_nested || translation_report.has_field_labels {
+            prefilter_queries.push(translation_report.expression);
+        } else {
+            search_queries.push(translation_report.expression);
+        }
+    }
+
+    Ok(QueryAnalysis {
+        prefilter_query: if prefilter_queries.len() <= 1 {
+            prefilter_queries.pop()
+        } else {
+            Some(BooleanExpression::Operation(BooleanOperation {
+                operator: Operator::And,
+                operands: prefilter_queries,
+            }))
+        },
+
+        search_query: if search_queries.len() <= 1 {
+            search_queries.pop()
+        } else {
+            Some(BooleanExpression::Operation(BooleanOperation {
+                operator: Operator::And,
+                operands: search_queries,
+            }))
+        },
+    })
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I would like a test or two hitting `translate()`. All the smaller functions are covered, but no test really covers `translate`.